### DISCUSSION TITLE
Text preprocessing

### DISCUSSION THREAD

**Discussion by chunibyocola** [chunibyocola/sc-translator-crx#17] [Show and tell]
## Overview
There are two types of `Text preprocessing`, `Before sending translate request` and  `After selecting text`.

This feature is potentially destructive. If you are not familiar with regular expressions, you are not recommended to write your own regular expressions.

**Feel free** to leave a comment below to ask for help or share your regular expressions.
## Before sending translate request
The text you want to translate will be processed before sending a translate request.

With this, you can remove the `Code annotation`, remove the `Extra spaces`, etc.
## After selecting text
The text you selected in the webpage will be processed.

With this, you can make this extension to ignore the text of some specific languages (not showing buttons and not translating automatically).
## Implementation
This feature implement by:
```javascript
text = text.replace(new RegExp(pattern, flags), replacement));
```
There are some useful requirements that cannot be achieved under this Implementation, such as: convert camel case. Considering this situation, a `Preset` process is provided.
## Preset
There is a `Preset` section showing what kind of `Preset` is provided.

Like `Convert camel case` option, enabling this option, you can make a text `aCamelCaseText` convert to `a camel case text`.
## Execute in order
The different execute in order will lead to different results.

For example, there are two rules `pattern: "\s+", flags: "g", replacement: " "` and `pattern: "//", flags: "g", replacement: ""` work on:
```
// A example
// to show
// the difference.
```
If the first rule is executed first, the result will be `A example  to show  the difference.`.
As you can see there is extra space between `example` and `to`, `show` and `the`. It is not good enough, even though it works too.

If the second rule is executed first, the result will be `A example to show the difference.`. There is no extra space.
## In conclution
This feature is very useful in some cases. But there are also some potential risks. Without it, extension also works well.

**Comment by chunibyocola** [comment]
## Rule share (Before sending translate request)
**pattern**: "//"
**flags**: "g"
**replacement**: ""

It is very useful to translate code comment.
It can convert `// example` to ` example`.

**Comment by chunibyocola** [comment]
## Rule share (Before sending translate request)
**pattern**: "\s+"
**flags**: "g"
**replacement**: " "

It is very useful to the text that contains mutiple spaces and line breaks.
It can convert
```
Here 
is some 
text
```
to `Here is some text`.

**Comment by chunibyocola** [comment]
## Rule share (After selecting text)
**pattern**: "[\u4e00-\u9fa5]"
**flags**: "g"
**replacement**: ""

This can make extension ignore the text you select with pure Chinese.
It can convert `这里是一段中文` to ` `.

**Comment by ecybron** [comment]
Hi,
I apologize if this question seems basic. I am new to this field and would appreciate your guidance.
I noticed that after translating the text, the punctuation marks that were after the code annotation appeared before it. How can this be fixed?

Code: `background color to <code>green</code>, as seen here.`
Before: "..background color to green, as seen here..."
After: "..цвет фона на , greenкак показано здесь..."

Bing/ En - Ru
Source: https://learn.shayhowe.com/html-css/getting-to-know-css/

**Reply by chunibyocola** [reply]
It seems you are using page translation on this page that text preprocessing is not working on.

And this issue can not be fixed, because the result is coming from remote server not local. I tried Google's, it does display the comma after `green`.

I tried to use offical's to work on the page, it shows the same result.
